### PR TITLE
As a user, I can save and resume a game

### PR DESCRIPTION
**Pull request by ellehallal**
GameSaver:
- create class: saves an instance of `Game` to a YAML file
- add `save` method to save to file
- add `retrieve_file_contents` to load file contents

GameLoader:
- create class: loads an instance of `Game` from a YAML file
- add `load` method to return requested object from file
- add `retrieve_file_contents` to load file contents
- `existing_game_names` to create array of existing game names

GameValidator:
- create class: validates user's input when loading game type
- `validate_game_type_input` checks if user enters `'new'` or `'existing'`
- `new_game_name` returns `game_name` if it is not present in the file
- `existing_game_selection` returns `game_name` if it is present in the file
- Plus other private methods

GameFactory:
- amend to initialise with instances of GameValidator, GameLoader and the file path
- `existing_game` method validates `game_name`, and returns the requested saved game
- amend `create_game` to either create a new game or load an existing game

HumanPlayer:
- `save_or_exit?` returns true if user input contains `'save'` or `'exit'`
- `handle_move` is extracted from `choose_move`. Validates if selected move is available on board
-  amend `choose_move`

Game:
- add instance variables `@save_game` and '@exit_game`
- add `save?`, `exit?` and `reset_save_game` methods
- amend `play_move`

Controller:
- add `save_game?` - if `game.save_game` is true, it saves the instance of the game
- add `exit_game?` - exits the app if `game.exit_game` is true
- add `goodbye` - displays exit message and exits the app
- amend `main_game`

Display:
- add `game_type_prompt`, `invalid_game_type_message`, `game_name_exists_message`, `save_exit_message`, `save_game_confirmation`, `create existing_game_names`

### CODE REVIEW

File: lib/game.rb [hunk 1]
```diff
@@ -1,18 +1,27 @@
...
-  attr_reader :board, :current_player, :opponent
+  attr_reader :board, :current_player, :opponent, :save_game, :exit_game
 
   def initialize(board, player1, player2)
     @board = board
     @player1 = player1
     @player2 = player2
     @current_player = @player1
     @opponent = @player2
+    @save_game = false
```
Comment (Kotauror) [inline lib/game.rb:10]: I was wondering what is your opinion on having primitive types as statuses (`save_game` and `exit_game`). 

For example, if in the future you were requested to implement exiting game with different statuses (like you do with `0` right now), you couldn't store this information in `@exit_game` status as it's only a primitive value.

File: lib/game_validator.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
+require 'yaml'
+
+class GameValidator
+  def initialize(display, game_loader)
+    @display = display
+    @game_loader = game_loader
+  end
+
+  def validate_game_type_input(filename)
+    if existing_games_empty?(filename)
```
Comment (Kotauror) [inline lib/game_validator.rb:10]: You could use a tenary operator to make it shorter.

File: lib/game_validator.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    @display.existing_game_names(existing_games)
+
+    game_name = prompt_for_name
+    until game_name_exists?(filename, game_name)
+      game_name = prompt_for_correct_name
+    end
+    game_name
+  end
+
+  private
```
Comment (Kotauror) [inline lib/game_validator.rb:36]: The `private` part of this class is awesome! 👏 Short, easy to understand methods having single responsibilities, good job!

File: lib/game_validator.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+  def retrieve_game_type
```
Comment (Kotauror) [inline lib/game_validator.rb:38]: why not `get_game_type`? (non-native speaker here :P)

File: lib/human_player.rb [hunk 5]
```diff
@@ -1,12 +1,28 @@
...
-  def choose_move(board, *args)
+  def choose_move(board, *_args)
     @display.display_board(board.squares)
     @display.show_current_player(mark, name)
     @display.ask_for_move
 
-    move = $stdin.gets.chomp.to_i
+    move = $stdin.gets.chomp
+
+    if save_or_exit?(move)
```
Comment (Kotauror) [inline lib/human_player.rb:11]: Did you try tenary operator?

File: spec/game_loader_spec.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
+require 'yaml'
+require 'game_loader'
+require_relative './test_doubles/fake_class_double'
+
+RSpec.describe GameLoader do
+  describe 'Loading an object' do
```
Comment (Kotauror) [inline spec/game_loader_spec.rb:6]: You could be more specific - `an object` doesn't tell much to a reader.

File: spec/game_loader_spec.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    end
+  end
+
+  describe 'Retrieving file contents' do
+    it 'returns file contents as a hash' do
+      game_loader = GameLoader.new
+      filename = './spec/test_data/game_loader_test.yml'
+
+      file = game_loader.retrieve_file_contents(filename)
+      p file
```
Comment (Kotauror) [inline spec/game_loader_spec.rb:38]: Is this `p` necessary?

File: spec/game_saver_spec.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  File.open(filename, 'w') { |file| file.truncate(0) }
+end
+
+RSpec.describe GameSaver do
+  describe 'Saving an object' do
+    it 'saves a hash, with the key "TicTacToe" and value "text"' do
+      game_saver = GameSaver.new
+      filename = './spec/test_data/test.yml'
+      game_name = 'TicTacToe'
+      obj = 'text'
```
Comment (Kotauror) [inline spec/game_saver_spec.rb:21]: Don't use `obj` or `string` etc. as names of variables as they, well, don't tell much about the role of the variable.

File: spec/game_saver_spec.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+      game_saver.save(filename, game_name, obj)
+      open_file = YAML.load_file(filename)
+
+      expect(open_file['TicTacToe']).to eq('text')
+
+      clear_file(filename)
```
Comment (Kotauror) [inline spec/game_saver_spec.rb:28]: Good job cleaning the file after the test!

File: spec/game_validator_spec.rb [hunk 10]
```diff
@@ -0,0 +1,122 @@
+require_relative './test_doubles/display_colour_double'
+require_relative './test_doubles/fake_class_double'
+require 'game_validator'
+require 'game_loader'
+require 'display'
+require 'yaml'
+
+def file_setup
+  filename = './spec/test_data/test.yml'
+  game_obj = 'game_object'
```
Comment (Kotauror) [inline spec/game_validator_spec.rb:10]: Same as above, the names of variables shouldn't have their types in names eg. `game_string` or `game_obj`. If you want to indicate what the variable holds, you could say `game_data`.

File: spec/test_doubles/fake_class_double.rb [hunk 11]
```diff
@@ -0,0 +1,9 @@
+class FakeClassDouble
```
Comment (Kotauror) [inline spec/test_doubles/fake_class_double.rb:1]: The name of the class is very mysterious... Can you think of a better one?

**Comment by Kotauror** [comment]
I think that a good practice for the future would be to slightly change the way you make commits. Instead of commits that add single methods eg. https://github.com/itsellej/ruby-tic-tac-toe/pull/20/commits/970167738b9b2ec5750902d21cb8da3641a28eb0 you could think about making commits that contain a part of business logic eg. one commit for saving the game, one commit for loading the game etc. 

If you missed a detail and you realise it needs to be added, like in this commit - https://github.com/itsellej/ruby-tic-tac-toe/pull/20/commits/704b25713e47d5d3d54dadf239b0fa5359f5bae8 - I would suggest trying to add it to already existing commit that is related to it. It would make the commit history cleaner and PRs easier to review.

On the other hand, I appreciate a lot the effort you put into the commit messages - they are neat!

File: lib/game_validator.rb (hunk 2 above)
Comment (ellehallal) [reply lib/game_validator.rb:10]: Good point, thank you

File: lib/game_validator.rb (hunk 3 above)
Comment (ellehallal) [reply lib/game_validator.rb:36]: Thanks :)

File: lib/game_validator.rb (hunk 4 above)
Comment (ellehallal) [reply lib/game_validator.rb:38]: I agree. `get_game_type` is clearer than `retrieve_game_type`. I'll change this, thanks

File: lib/human_player.rb (hunk 5 above)
Comment (ellehallal) [reply lib/human_player.rb:11]: I did not, but I will now :)

File: spec/game_loader_spec.rb (hunk 6 above)
Comment (ellehallal) [reply spec/game_loader_spec.rb:6]: I agree. I'll change this, thanks!

File: spec/game_loader_spec.rb (hunk 7 above)
Comment (ellehallal) [reply spec/game_loader_spec.rb:38]: It isn't. I'll remove it

File: spec/game_saver_spec.rb (hunk 8 above)
Comment (ellehallal) [reply spec/game_saver_spec.rb:21]: Ok, I'll change this. Thanks

File: spec/game_saver_spec.rb (hunk 9 above)
Comment (ellehallal) [reply spec/game_saver_spec.rb:28]: Thanks!

File: spec/test_doubles/fake_class_double.rb (hunk 11 above)
Comment (ellehallal) [reply spec/test_doubles/fake_class_double.rb:1]: Good point, I'll change this

File: lib/controller.rb [hunk 12]
```diff
@@ -1,21 +1,47 @@
...
   end
 
   def main_game
     game_setup
-    play_move until @game.over?
+    until @game.over?
+      play_move
+      save_game?
+      exit_game?
+    end
```
Comment (matthewglover) [inline lib/controller.rb:18]: This reads really nicely :)

File: lib/controller.rb [hunk 13]
```diff
@@ -1,21 +1,47 @@
...
+  end
+
+  def exit_game?
+    goodbye if @game.exit_game
+  end
+
+  def goodbye
+    @display.show_exit_message
+    exit(0)
```
Comment (matthewglover) [inline lib/controller.rb:39]: Could this use the game loop to exit? It adds complexity to have a game loop that can be exited directly in a method call within the game loop. At this stage it is manageable, but imagine if there were more ways to exit the game loop - how easy would it be to follow all possible combinations?

File: lib/controller.rb (hunk 12 above)
Comment (ellehallal) [reply lib/controller.rb:18]: Thanks!

File: lib/controller.rb (hunk 13 above)
Comment (ellehallal) [reply lib/controller.rb:39]: Good point. I'll look into this, thanks :)

**Pull request merged**